servlet/src/main/java/io/undertow/servlet/spec/ServletOutputStreamImpl.java

```diff
@@ -541,35 +546,31 @@ public void transferFrom(FileChannel source) throws IOException {
             if (channel == null) {
                 channel = servletRequestContext.getExchange().getResponseChannel();
             }
-            long position = source.position();
-            long count = source.size() - position;
-            Channels.transferBlocking(channel, source, position, count);
+            Channels.transferBlocking(channel, source, startPosition, count);
             updateWritten(count);
         } else {
             setFlags(FLAG_WRITE_STARTED);
             createChannel();
 
             long pos = 0;
             try {
-                long size = source.size();
-                pos = source.position();
+                long remaining = count;
+                pos = startPosition;
 
-                while (size - pos > 0) {
-                    long ret = channel.transferFrom(pendingFile, pos, size - pos);
+                while (remaining - pos > 0) {
+                    long ret = channel.transferFrom(pendingFile, pos, remaining - pos);
                     if (ret <= 0) {
                         clearFlags(FLAG_READY);
                         pendingFile = source;
-                        source.position(pos);
```

Comment:
this is the piece that seems incongruous

Reply:
yes, I'm not sure why it is doing that. Also, notice that this could throw a NPE, because pendingFile has not been assigned to source